### PR TITLE
Feature/rq dashboard

### PR DESCRIPTION
**Pull request by albrja**
### Description
Added functionality to run a RQ-dashboard when user launches a parallel simulation.
- *Category*: Feature
- *JIRA issue*: [MIC-2934](https://jira.ihme.washington.edu/browse/MIC-2934)

-Added run_rq_dashboard function that launches dashboard during main psimulate runner loop
-Dashboard captures all jobs submitted on that host for all redis databases that were provide when dashboard was initialized (current implementation is all redis databases but this could be changed with little work).
-Hook to kill subprocess running dashboard when the parallel simulation ends.
-Added rq.log to log subprocess and provide user with URL to navigate to use the dashboard.  URL is also printed to terminal.

### Testing
Ran multiple parallel simulation and successfully saw working implementation for:
    -Creation and usage of new rq.log file
    -Dashboard launches at start of simulation and the process ends when the simulation finishes
    -All redis databases used for parallel simulation are attached to the dashboard depending on how many were used for that simulation.

### CODE REVIEW

File: src/vivarium_cluster_tools/psimulate/runner.py [hunk 1]
```diff
@@ -107,6 +101,26 @@ def try_run_vipin(log_path: Path) -> None:
...
+    split_urls = " -u ".join(url for url in redis_urls)
+    command = "rq-dashboard -u " + split_urls + " --debug"
+
+    rq_dashboard_log.write(f"Dashboard running at http://{hostname}:9181\n")
+    logger.info(f"Dashboard running at http://{hostname}:9181")
+    proc = subprocess.Popen(
+        command, shell=True, stdout=rq_dashboard_log, stderr=rq_dashboard_log
+    )
+
+    atexit.register(proc.kill)
```
Comment (albrja) [inline src/vivarium_cluster_tools/psimulate/runner.py:121]: Hook to kill dashboard when psimulate runner loop ends.

File: src/vivarium_cluster_tools/psimulate/runner.py [hunk 2]
```diff
@@ -107,6 +101,26 @@ def try_run_vipin(log_path: Path) -> None:
...
+    # Get generic hostname url
+    url = redis_urls[0].split("//")[1]
+    hostname = url.split(":")[0]
+
+    # Set up log file
+    rq_dashboard_log = (output_directory / "rq.log").open("a")
+    logger.info("Fetching redis urls and starting RQ-Dashboard")
+    split_urls = " -u ".join(url for url in redis_urls)
+    command = "rq-dashboard -u " + split_urls + " --debug"
+
```
Comment (albrja) [inline src/vivarium_cluster_tools/psimulate/runner.py:114]: I think this implementation could be improved depending on what we want.  Currently, all redis urls are submitted to the dashboard but it doesn't have to be that way.  Users could also open up another dashboard with just one redis db.  There are a lot of options here.

File: src/vivarium_cluster_tools/psimulate/runner.py [hunk 3]
```diff
@@ -107,6 +101,26 @@ def try_run_vipin(log_path: Path) -> None:
...
+    rq_dashboard_log.write(f"Dashboard running at http://{hostname}:9181\n")
```
Comment (albrja) [inline src/vivarium_cluster_tools/psimulate/runner.py:115]: This log currently only captures the local url and the url that the user can navigate to for the dashboard.  Room for improvement here.

File: src/vivarium_cluster_tools/psimulate/runner.py [hunk 4]
```diff
@@ -48,10 +49,7 @@ def process_job_results(
 
             if len(unwritten_results) > batch_size:
                 written_results, unwritten_results = results.write_results_batch(
-                    output_directory,
-                    written_results,
-                    unwritten_results,
-                    batch_size,
+                    output_directory, written_results, unwritten_results, batch_size
```
Comment (collijk) [inline src/vivarium_cluster_tools/psimulate/runner.py:52]: Did you delete the trailing comma here?  That's what induces the whitespace noise.

File: src/vivarium_cluster_tools/psimulate/runner.py [hunk 5]
```diff
@@ -107,6 +102,26 @@ def try_run_vipin(log_path: Path) -> None:
         logger.warning(f"Performance reporting failed with: {e}")
 
 
+def run_rq_dashboard(redis_urls: list, output_directory: Path) -> None:
```
Comment (collijk) [inline src/vivarium_cluster_tools/psimulate/runner.py:105]: I'd make a `monitoring` module and put this function in it. It'll eventually become a subpackage with a bunch of other stuff in it.

File: src/vivarium_cluster_tools/psimulate/runner.py (hunk 2 above)
Comment (collijk) [reply src/vivarium_cluster_tools/psimulate/runner.py:115]: We want all redis urls in the same dashboard, we just want to change the way the dashboard handles the information

File: src/vivarium_cluster_tools/psimulate/runner.py [hunk 6]
```diff
@@ -107,6 +102,26 @@ def try_run_vipin(log_path: Path) -> None:
...
+    # Get generic hostname url
+    url = redis_urls[0].split("//")[1]
+    hostname = url.split(":")[0]
+
+    # Set up log file
+    rq_dashboard_log = (output_directory / "rq.log").open("a")
```
Comment (collijk) [inline src/vivarium_cluster_tools/psimulate/runner.py:111]: `rq.log` is a bad name for this.  Use `rq_dashboard.log`.

File: src/vivarium_cluster_tools/psimulate/runner.py [hunk 7]
```diff
@@ -107,6 +102,26 @@ def try_run_vipin(log_path: Path) -> None:
...
+    logger.info("Fetching redis urls and starting RQ-Dashboard")
+    split_urls = " -u ".join(url for url in redis_urls)
+    command = "rq-dashboard -u " + split_urls + " --debug"
+
+    rq_dashboard_log.write(f"Dashboard running at http://{hostname}:9181\n")
```
Comment (collijk) [inline src/vivarium_cluster_tools/psimulate/runner.py:116]: This is a lie.  YOu haven't launched the dashboard at the time you're writing to this log file.

File: src/vivarium_cluster_tools/psimulate/runner.py [hunk 8]
```diff
@@ -107,6 +102,26 @@ def try_run_vipin(log_path: Path) -> None:
...
+    logger.info(f"Dashboard running at http://{hostname}:9181")
```
Comment (collijk) [inline src/vivarium_cluster_tools/psimulate/runner.py:117]: As is this.

File: src/vivarium_cluster_tools/psimulate/runner.py (hunk 5 above)
Comment (collijk) [reply src/vivarium_cluster_tools/psimulate/runner.py:105]: Also, call your `output_directory` something more descriptive like `logging_root`.

File: src/vivarium_cluster_tools/psimulate/runner.py [hunk 9]
```diff
@@ -107,6 +102,26 @@ def try_run_vipin(log_path: Path) -> None:
...
 
+def run_rq_dashboard(redis_urls: list, output_directory: Path) -> None:
+    # Get generic hostname url
+    url = redis_urls[0].split("//")[1]
+    hostname = url.split(":")[0]
+
+    # Set up log file
+    rq_dashboard_log = (output_directory / "rq.log").open("a")
+    logger.info("Fetching redis urls and starting RQ-Dashboard")
+    split_urls = " -u ".join(url for url in redis_urls)
```
Comment (collijk) [inline src/vivarium_cluster_tools/psimulate/runner.py:113]: Two things:

`" -u ".join(url for url in redis_urls)` is the same as `" -u ".join(redis_urls)`

and 
`url_flags = " ".join([f'-u {url}' for url in redis_urls])`
means you can write
`command = "rq-dashboard " + url_flags + " --debug"`
without the fencepost `-u` in the command.

File: src/vivarium_cluster_tools/psimulate/runner.py [hunk 10]
```diff
@@ -208,6 +223,15 @@ def main(
     # Generate a worker template that chooses a redis DB at random to connect to.
     # This should (approximately) evenly distribute the workers over the work.
     redis_urls = [f"redis://{hostname}:{port}" for hostname, port in redis_ports]
+
+    # Grab redis urls for dashboard and send them to function for popen
+    rq_urls = [
+        f"redis://{hostname}.cluster.ihme.washington.edu:{port}"
```
Comment (collijk) [inline src/vivarium_cluster_tools/psimulate/runner.py:229]: The workers are okay with this change?

File: src/vivarium_cluster_tools/psimulate/runner.py (hunk 5 above)
Comment (albrja) [reply src/vivarium_cluster_tools/psimulate/runner.py:105]: Ok I will do this.

File: src/vivarium_cluster_tools/psimulate/runner.py (hunk 4 above)
Comment (albrja) [reply src/vivarium_cluster_tools/psimulate/runner.py:52]: I've been having issues with black/isort so I'll add this back in.

File: src/vivarium_cluster_tools/psimulate/runner.py (hunk 10 above)
Comment (albrja) [reply src/vivarium_cluster_tools/psimulate/runner.py:229]: I'm not sure I understand your question but this isn't doing anything to the workers it is just formatting the redis urls the way the rq-dashboard wants them.   It isn't doing anything to the workers themselves to my knowledge.

**Comment by albrja** [comment]
One important note is I had to make a few changes to the cli.py file in the rq-dashboard package since the package itself has not been updated in a few years but click has so to get this working I had to make my own branch to fix some default settings for click options which is not reflected in this PR and the RQ-Dashboard as implemented here will not work without making these changes in the rq-dashboard package.

File: src/vivarium_cluster_tools/psimulate/runner.py (hunk 10 above)
Comment (mattkappel) [reply src/vivarium_cluster_tools/psimulate/runner.py:209]: So it's changing the workers by giving them the FQD of `.cluster.ihme.washington.edu`. I would presume that so long as the cluster's FQD doesn't change, we're ok.

Comment (stevebachmeier) [reply src/vivarium_cluster_tools/psimulate/runner.py:209]: @mattkappel Are you saying that f"redis://{hostname}:{port}" resolves to "redis://{hostname}.cluster.ihme.washington.edu:{port}"? If so, then no need for an rq_urls and you can just pass in redis_urls?

File: src/vivarium_cluster_tools/psimulate/monitoring.py [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    hostname = url.split(":")[0]
+
+    # Set up log file
+    rq_dashboard_log = (logging_root / "rq_dashboard.log").open("a")
+    logger.info("Fetching redis urls and starting RQ-Dashboard")
+    url_flags = " ".join([f"-u {url}" for url in redis_urls])
+    command = "rq-dashboard " + url_flags + " --debug"
+
+    proc = subprocess.Popen(
+        command, shell=True, stdout=rq_dashboard_log, stderr=rq_dashboard_log
```
Comment (stevebachmeier) [inline src/vivarium_cluster_tools/psimulate/monitoring.py:25]: Do you want stdout and stderr going to the same log?

File: src/vivarium_cluster_tools/psimulate/monitoring.py [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+    )
+    rq_dashboard_log.write(f"Dashboard running at http://{hostname}:9181\n")
+    logger.info(f"Dashboard running at http://{hostname}:9181")
```
Comment (stevebachmeier) [inline src/vivarium_cluster_tools/psimulate/monitoring.py:28]: Where did these static ports come from?

File: src/vivarium_cluster_tools/psimulate/runner.py (hunk 10 above)
Comment (mattkappel) [reply src/vivarium_cluster_tools/psimulate/runner.py:209]: oh sorry, I misread! `rq_urls != redis_urls` But I would think what you think I said is true.

**Review by rmudambi — APPROVED** [review]

File: src/vivarium_cluster_tools/psimulate/monitoring.py (hunk 11 above)
Comment (albrja) [reply src/vivarium_cluster_tools/psimulate/monitoring.py:25]: Yes.

File: src/vivarium_cluster_tools/psimulate/monitoring.py (hunk 12 above)
Comment (albrja) [reply src/vivarium_cluster_tools/psimulate/monitoring.py:28]: In the vivarium_dashboard package the cli.py that has all the click options sets 9181 as the default.

File: src/vivarium_cluster_tools/psimulate/monitoring.py (hunk 11 above)
Comment (mattkappel) [reply src/vivarium_cluster_tools/psimulate/monitoring.py:25]: I mean, it's sometimes useful to have an `log.o` and and `log.e` but generally I don't think that's useful for our logging purposes in `psimulate`.

File: src/vivarium_cluster_tools/psimulate/monitoring.py (hunk 12 above)
Comment (mattkappel) [reply src/vivarium_cluster_tools/psimulate/monitoring.py:28]: We ought to have psimulate pick a random open port like @collijk suggested and pass that into as an option.

**Pull request closed without merging**